### PR TITLE
Switch server release group to pnpm

### PR DESCRIPTION
**Pull request by tylerbutler**
This PR updates the server release group to use pnpm instead of npm. Summary of changes:

- lerna was bumped to v5 in the server release group. This was needed because earlier releases only supported yarn workspaces properly. I've attempted this upgrade without pnpm in the past, but it always failed (presumably due to npm bugs).
- lerna uses the pnpm workspace config and uses pnpm to execute scripts.
- Several packages had missing deps on node libraries like `assert` and `events`, so I added those explicitly.
- Some packages had missing non-node deps, like axios and sinon. My guess is that these worked because of hoisting in the past.
- The Docker image uses debian buster instead of stretch. This was needed to get a later release of the Python build tools that the server packages require to build (due to node-rdkafka). The latest releases available for stretch are too old. But if we must continue to use stretch, we can manually build the tools in the image if we have to. I'd prefer to use a later debian base. 🤞

## Notes for reviewers

I am particularly nervous about the following areas:

- Dockerfile changes.
- winston and express-related dependency updates. The typing of these packages is quite strange to me, so I worry that I broke something.
- Should the assert polyfill be used, or should I add a dep on the isomorphic implementation in common-utils?
- I'm considering bumping the version to the next major in case this introduces breaking changes. Thoughts?

### CODE REVIEW

**Comment by markfields** [comment]
> Should the assert polyfill be used, or should I add a dep on the isomorphic implementation in common-utils?

Why would this PR be concerned with this? i.e. what changes?

**Comment by tylerbutler** [comment]
> > Should the assert polyfill be used, or should I add a dep on the isomorphic implementation in common-utils?
> 
> Why would this PR be concerned with this? i.e. what changes?

Several packages have `import assert from "assert";` in their source files (not test code). There are also two tests that attempt to webpack some of the packages, presumably to ensure that we don't add deps that require special handling in bundler config. Example:

https://github.com/microsoft/FluidFramework/blob/50fdacffccbcda0ace7cfd2431ca0d373aecc8a0/server/routerlicious/packages/lambdas/src/test/webpack.spec.ts#L10-L12

Options to address it are (1) adding `assert` and/or `events` dependencies explicitly to the packages that use those modules, OR (2) update the webpack config to ignore those polyfills. (1) is what the PR does right now, because I think it's more correct. But option 2 has been used before, it seems:

https://github.com/microsoft/FluidFramework/blob/50fdacffccbcda0ace7cfd2431ca0d373aecc8a0/server/routerlicious/packages/lambdas/src/test/webpack.spec.ts#L37-L40

As for what changes, I think it depends on the consumer of the package, and whether they're bundling for the browser. If they are, then without these changes the bundler _must_ be configured to load the polyfills.

**Review by hedasilv — COMMENTED** [review]
Hey @tylerbutler  - I cloned your fork and checked out the `pnpm/server` branch to run some tests, as we had discussed. I noticed a few things:

* `flub` issues - I had to run `npm install -g flub` and also run `npm install` from root. Not sure if these `flub` issues are related to this change or not.
* `npm install` does not work anymore under server/routerlicious. The error message says we need to use `pnpm`. We need to update documentation to tell users to install `pnpm` (can be done with `npm install -g pnpm`)
* server/routerlicious/package.json had many scripts based on `npm`. Do those also need to be updated to `pnpm`? Or would things work if we keep using `npm`, such as `npm run build` ([link](https://github.com/microsoft/FluidFramework/blob/23f4f50cd42818a2f135f7418355f366b36a5008/server/routerlicious/package.json#L14))?
* This change is labeled as "switch server release  group to pnpm", but it only touches routerlicious. Do the same changes also need to be applied to server/historian, server/gitrest, server/tinylicious, etc?

**Comment by tylerbutler** [comment]
> npm install does not work anymore under server/routerlicious. The error message says we need to use pnpm. We need to update documentation to tell users to install pnpm (can be done with npm install -g pnpm)

You should have gotten an error message saying to do use npm -- did you not get that message?

> server/routerlicious/package.json had many scripts based on npm. Do those also need to be updated to pnpm?

No need to update them; npm works fine to run scripts in most cases.

> This change is labeled as "switch server release group to pnpm", but it only touches routerlicious. Do the same changes also need to be applied to server/historian, server/gitrest, server/tinylicious, etc?

The server release group is only r11s. Release groups can be moved individually to npm (or yarn) as needed, but it's not urgent. We deliberately support both npm/lerna and pnpm in fluid-build so we can move gradually. That said, once people get a taste of the perf improvements there might be some more PRs coming. :smile:

**Comment by tylerbutler** [comment]
> flub issues - I had to run npm install -g flub and also run npm install from root. Not sure if these flub issues are related to this change or not.

Do you remember when this happened? This shouldn't be needed.

**Comment by hedasilv** [comment]
@tylerbutler , Got it, thanks for all the info!

> You should have gotten an error message saying to use pnpm -- did you not get that message?

I did get the pnpm error message when I tried to run `npm install`. It was just not clear to me that this change would require that to be different.

> No need to update them; npm works fine to run scripts in most cases.

I see. So I guess this means that: for `npm install`, we now need to use `pnpm install` instead. But we can keep everything else, including scripts based on `npm`. Is that correct?

> The server release group is only r11s. Release groups can be moved individually to npm (or yarn) as needed, but it's not urgent. We deliberately support both npm/lerna and pnpm in fluid-build so we can move gradually. That said, once people get a taste of the perf improvements there might be some more PRs coming. 😄

I see. I think my point of confusion is that I don't understand what "The server release group is only r11s" means. What is this server release group? Does release here mean Docker image release? Or npm package release? Also, since the server depends on Historian and GitRest, don't those have to be released as well somehow?



> > flub issues - I had to run npm install -g flub and also run npm install from root. Not sure if these flub issues are related to this change or not.
> 
> Do you remember when this happened? This shouldn't be needed.

It was immediately after I cloned the repo. After cloning the repo - and before checking out the `pnpm/server` branch, so I was on main - I went to the server/routerlicious path, and ran `npm install`. All worked. Then, I ran `npm run build`. That gave me an error that flub was not a recognized command/executable. I fixed that by doing `npm install -g flub`, which is weird, because if it's something required for policy check, I already had it globally installed due to my personal FluidFramework repo fork. Then, after I got past that error message, I retried `npm run build` again, under `server/routerlicious`. Another flub error happened, but this time saying that some particular files could not be found. Unfortunately I can't seem to find the terminal where I was running the commands to provide the actual error messages. This last error was only solved after I ran `npm install` from the repo's root. After that, running `npm run build` on `server/routerlicious` succeeded.

Regarding testing:
Thanks for the clarifications above. I resumed testing. From server/routerlicious, I did:
1. pnpm install
2. npm run build
3. npm run docker-build
4. npm start

All services/containers that talk to Kafka started failing with errors like this:

> routerlicious-scriptorium-1  | error: Error: /usr/src/server/node_modules/.pnpm/zookeeper@5.6.0/node_modules/zookeeper/build/Release/zookeeper.node: invalid ELF header
routerlicious-scriptorium-1  |     at Object.Module._extensions..node (internal/modules/cjs/loader.js:1144:18)
routerlicious-scriptorium-1  |     at Module.load (internal/modules/cjs/loader.js:950:32)
routerlicious-scriptorium-1  |     at Function.Module._load (internal/modules/cjs/loader.js:790:12)
routerlicious-scriptorium-1  |     at Module.require (internal/modules/cjs/loader.js:974:19)
routerlicious-scriptorium-1  |     at require (internal/modules/cjs/helpers.js:101:18)
routerlicious-scriptorium-1  |     at load (/usr/src/server/node_modules/.pnpm/node-gyp-build@4.5.0/node_modules/node-gyp-build/index.js:22:10)
routerlicious-scriptorium-1  |     at Object.<anonymous> (/usr/src/server/node_modules/.pnpm/zookeeper@5.6.0/node_modules/zookeeper/lib/zookeeper.js:7:43)
routerlicious-scriptorium-1  |     at Module._compile (internal/modules/cjs/loader.js:1085:14)
routerlicious-scriptorium-1  |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
routerlicious-scriptorium-1  |     at Module.load (internal/modules/cjs/loader.js:950:32) {
routerlicious-scriptorium-1  |   code: 'ERR_DLOPEN_FAILED'
routerlicious-scriptorium-1  | } {"label":"winston","messageMetaData":{"documentId":"","tenantId":""},"timestamp":"2023-01-13T19:40:29.489Z"}

That usually happens when we build packages that have native code (I think node-rdkafka, nodegit and zookeeper are examples or packages implemented in C++ or similar but which are wrapped up in JavaScript). [The package is built on an architecture different than the one used in Docker.](https://stackoverflow.com/a/29994851) We on the FRS team use Mac for development, and Docker is running Linux. That is why, in our docker-compose.dev.yaml, we add some volume mounting rules to prevent that the those packages built for the host architecture be mounted on Docker - for example, see:
https://github.com/microsoft/FluidFramework/blob/c1dd97009c6ae5bcd2f2f4558e23e02975943ee4/server/routerlicious/docker-compose.dev.yml#L6-L8

I think we might need to add new rules now that the path of the modules has changed: for example, in the error above I see .pnpm is part of the path:

>  /usr/src/server/node_modules/.pnpm/zookeeper@5.6.0/node_modules/zookeeper/build/Release/zookeeper.node

Now, regarding the other questions in the PR description:
1. Dockerfile changes -> those look good to me.
2. winston, express dependency updates: what did you mean by strange typing? When running the test above, express and winston seem to have worked (since I got logs, and the HTTP server received my Clicker client's requests).
3. assert polyfill  -> I am not sure.
4. Bumping major version -> I think that would be ok from our perspective.

**Comment by tylerbutler** [comment]
> I see. So I guess this means that: for npm install, we now need to use pnpm install instead. But we can keep everything else, including scripts based on npm. Is that correct?

Correct. You can also use `pnpm run` but you don't have to.

> I see. So I guess this means that: for npm install, we now need to use pnpm install instead. But we can keep everything else, including scripts based on npm. Is that correct?

> I think my point of confusion is that I don't understand what "The server release group is only r11s" means. What is this server release group?

No worries! The release process is always changing. :smile: "Release group" is the term for a group of packages that we version and release together. Client is the primary release group, but there's also server, azure, and build-tools, plus independent packages that we release separately, like tinylicious or the eslint config. There's an ADO pipeline for each release group.

The gitrest and other server projects are not included in any release groups. This may not be correct, but since we don't release those projects from the client team, it's probably fine.

From a dev workflow perspective, release groups are interesting if you use fluid-build (aka `npm run build:fast`), because they define the groups of projects the fluid-build knows how to build incrementally.

> Also, since the server depends on Historian and GitRest, don't those have to be released as well somehow?

Each release group is separate, and dependencies between them are manually updated. So when you publish new historian packages you have to separately update the consumers to use the new package.

> It was immediately after I cloned the repo. After cloning the repo - and before checking out the pnpm/server branch, so I was on main - I went to the server/routerlicious path, and ran npm install.

Thanks! I'm pretty sure I understand the root cause of this, and it's not related to this change.

> I think we might need to add new rules now that the path of the modules has changed:

Good catch!

> winston, express dependency updates: what did you mean by strange typing? When running the test above, express and winston seem to have worked (since I got logs, and the HTTP server received my Clicker client's requests).

I'll comment on the code to point out what I mean. But it's grreat to know that the functionality you were expecting to work still does. 👍🏻

**Comment by hedasilv** [comment]
> > winston, express dependency updates: what did you mean by strange typing? When running the test above, express and winston seem to have worked (since I got logs, and the HTTP server received my Clicker client's requests).
> 
> I'll comment on the code to point out what I mean. But it's grreat to know that the functionality you were expecting to work still does. 👍🏻

Actually, I made a mistake in my comment above. I _assumed_ all was fine but then I started having the Kafka/ZooKeeper errors. Once the docker-compose.dev.yaml is updated, I will test things out again to confirm what I said above was true

File: server/routerlicious/package.json [hunk 1]
```diff
@@ -30,20 +30,13 @@
     "docker-build": "docker-compose build",
     "docker-clean": "docker-compose down --rmi local -v",
     "format": "lerna run format --parallel --stream",
-    "postinstall": "npm run postinstall:lerna",
+		"preinstall": "node ../../scripts/only-pnpm.cjs",
```
Comment (znewton) [inline server/routerlicious/package.json:33]: nit: Maybe this is just github, but the indent here looks way too long

**Review by znewton — APPROVED** [review]
Approving since I pulled this locally and validated that I could successfully run a fluid app against it

**Review by hedasilv — APPROVED** [review]

**Pull request merged**